internal/dbutil/sqliteutil/migrator.go

```diff
@@ -0,0 +1,183 @@
+	}
+
+	// Remove tables we don't want to keep from the cloned destination database.
+	err = KeepTables(ctx, destDB, tablesToKeep...)
+	if err != nil {
+		return errs.Wrap(err)
+	}
+	return nil
+}
+
+// backup executes the sqlite3 backup process that safely ensures that no other
+// connections to the database accidentally corrupt the source or destination.
+func backup(ctx context.Context, sourceDB *sqlite3.SQLiteConn, destDB *sqlite3.SQLiteConn) error {
+	// "main" represents the main (ie not "temp") database in sqlite3, which is
+	// the database we want to backup, and the appropriate dest in the destDB
+	backup, err := destDB.Backup("main", sourceDB, "main")
+	if err != nil {
+		return errs.Wrap(err)
+	}
+
+	isDone, err := backup.Step(0)
+	if err != nil {
+		return errs.Wrap(err)
+	}
+	if isDone {
+		return errs.New("Backup is done")
+	}
+
+	// Check that the page count and remaining values are reasonable.
+	initialPageCount := backup.PageCount()
+	if initialPageCount <= 0 {
+		return errs.New("initialPageCount invalid")
+	}
+	initialRemaining := backup.Remaining()
+	if initialRemaining <= 0 {
+		return errs.New("initialRemaining invalid")
+	}
+	if initialRemaining != initialPageCount {
+		return errs.New("initialRemaining != initialPageCount")
+	}
+
+	// Step -1 is used to copy the entire source database to the destination.
+	isDone, err = backup.Step(-1)
+	if err != nil {
+		return errs.Wrap(err)
+	}
+	if !isDone {
+		return errs.New("Backup not done")
+	}
+
+	// Check that the page count and remaining values are reasonable.
+	finalPageCount := backup.PageCount()
+	if finalPageCount != initialPageCount {
+		return errs.New("finalPageCount != initialPageCount")
+	}
+	finalRemaining := backup.Remaining()
+	if finalRemaining != 0 {
+		return errs.New("finalRemaining invalid")
+	}
+
+	// Finish the backup.
+	err = backup.Finish()
+	if err != nil {
+		return errs.Wrap(err)
+	}
+	return nil
+}
+
+// KeepTables drops all the tables except the specified tables to keep.
+func KeepTables(ctx context.Context, db *sql.DB, tablesToKeep ...string) error {
+	// Get a list of tables excluding sqlite3 system tables.
+	rows, err := db.Query("SELECT name FROM sqlite_master WHERE type ='table' AND name NOT LIKE 'sqlite_%';")
+	if err != nil {
+		return errs.Wrap(err)
+	}
+
+	// Collect a list of the tables. We must do this because we can't do DDL
+	// statements like drop tables while a query result is open.
+	var tables []string
+	for rows.Next() {
+		var tableName string
+		err = rows.Scan(&tableName)
+		if err != nil {
+			return errs.Wrap(err)
+		}
+		tables = append(tables, tableName)
+	}
+	err = rows.Close()
+	if err != nil {
+		return errs.Wrap(err)
+	}
+
+	// Loop over the list of tables and decide which ones to keep and which to drop.
+	for _, tableName := range tables {
+		if !tableToKeep(tableName, tablesToKeep) {
+			// Drop tables we aren't told to keep in the destination database.
+			_, err = db.Exec("DROP TABLE "+tableName+";", nil)
```

Comment:
why not `db.Exec("DROP TABLE ?", tableName)`? Regardless, passing `nil` here doesn't seem appropriate. That's passing `[]interface{}{nil}` rather than `[]interface{}{}` since the arguments list is variadic.

Reply:
Fixed.

Reply:
Sorry, I had to change it to `db.Exec(fmt.Sprintf("DROP TABLE %s;", tableName)`. It appears that sqlite does not allow table names to be used in parameter substitution.